execution/scheduler_ext_test.go

```diff
@@ -1045,96 +1048,92 @@ func TestDNSResolver(t *testing.T) {
+	testCases := map[string]struct {
+		opts              lib.Options
+		expFailedRequests uint32
+	}{
+		"default": { // IPs are cached for 5m
+			lib.Options{DNS: types.DefaultDNSConfig()}, 0,
+		},
+		"0": { // cache is disabled, every request does a DNS lookup
+			lib.Options{DNS: types.DNSConfig{
+				TTL:    null.StringFrom("0"),
+				Select: types.NullDNSSelect{DNSSelect: types.DNSfirst, Valid: true},
+				Policy: types.NullDNSPolicy{DNSPolicy: types.DNSpreferIPv4, Valid: false},
+			}}, 5,
+		},
+		"1000": { // cache IPs for 1s, check that unitless values are interpreted as ms
+			lib.Options{DNS: types.DNSConfig{
+				TTL:    null.StringFrom("1000"),
+				Select: types.NullDNSSelect{DNSSelect: types.DNSfirst, Valid: true},
+				Policy: types.NullDNSPolicy{DNSPolicy: types.DNSpreferIPv4, Valid: false},
+			}}, 4,
+		},
+		"3s": {
+			lib.Options{DNS: types.DNSConfig{
+				TTL:    null.StringFrom("3s"),
+				Select: types.NullDNSSelect{DNSSelect: types.DNSfirst, Valid: true},
+				Policy: types.NullDNSPolicy{DNSPolicy: types.DNSpreferIPv4, Valid: false},
+			}}, 3,
+		},
+	}
 
-				ctx, cancel, execScheduler, samples := newTestScheduler(t, runner, logger, tc.opts)
-				defer cancel()
+	for name, tc := range testCases {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
 
-				mr.Set("myhost", sr("HTTPBIN_IP"))
-				time.AfterFunc(1700*time.Millisecond, func() {
-					mr.Set("myhost", "127.0.0.254")
-				})
-				defer mr.Unset("myhost")
+			logger := logrus.New()
+			logger.SetOutput(io.Discard)
+			registry := metrics.NewRegistry()
+			builtinMetrics := metrics.RegisterBuiltinMetrics(registry)
+			runner, err := js.New(
+				&lib.TestPreInitState{
+					Logger:         logger,
+					BuiltinMetrics: builtinMetrics,
+					Registry:       registry,
+				},
+				&loader.SourceData{
+					URL: &url.URL{Path: "/script.js"}, Data: []byte(script),
+				}, nil)
+			require.NoError(t, err)
 
-				errCh := make(chan error, 1)
-				go func() { errCh <- execScheduler.Run(ctx, ctx, samples) }()
+			mr := mockresolver.New(nil, net.LookupIP)
+			runner.ActualResolver = mr.LookupIPAll
 
-				select {
-				case err := <-errCh:
-					require.NoError(t, err)
-					entries := logHook.Drain()
-					require.Len(t, entries, tc.expLogEntries)
-					for _, entry := range entries {
-						require.IsType(t, &url.Error{}, entry.Data["error"])
-						err, ok := entry.Data["error"].(*url.Error)
-						assert.True(t, ok)
-						errMsg := err.Error()
-						contains := strings.Contains(errMsg, "connection refused") ||
-							strings.Contains(errMsg, "request timeout") ||
-							strings.Contains(errMsg, "dial: i/o timeout")
-						assert.Truef(t, contains, "expected error to contain one of the list of messages")
+			var failedRequests uint32
+			samplesHandler := func(sc metrics.SampleContainer) {
+				if sc == nil {
+					return
+				}
+				for _, s := range sc.GetSamples() {
+					if s.Metric.Name == metrics.HTTPReqFailedName && s.Value == 1 {
```

Comment:
I was thinking about something like:
```suggestion
					if s.Metric.Name == metrics.HTTPReqsName && s.Value == 1 {
					    changeMockResolver <- struct{}{}
```
then

```go
go func() {
<-changeMockResolver
mr.Set("myhost", "127.0.0.254")
}()
```
I think it should remove the timer.After requirement

Reply:
I guess we could do that, but I don't like that we'd be relying on metric emission for the functionality of the test. That is, if `http_reqs` is never emitted, `mr.Set()` wouldn't be called, and the failure would be confusing.

And it doesn't solve the timing issue, since we still use `sleep()` in the script. But I'll give it a try, and also consider using `MiniRunner` instead.

Reply:
I removed checking samples in the latest force-push, so I'm not doing this anymore.